atc/db/versions_db.go

```diff
@@ -623,30 +590,73 @@ func (versions VersionsDB) migrateSingle(buildID int) (string, error) {
 					SELECT build_id, resource_id, json_agg(version_md5) AS v
 					FROM (
 						(
-							SELECT build_id, resource_id, version_md5 FROM build_resource_config_version_outputs o WHERE o.build_id = $1
+							SELECT build_id, resource_id, version_md5
+							FROM build_resource_config_version_outputs o
+							WHERE o.build_id = $1
 						)
 						UNION ALL
 						(
-							SELECT build_id, resource_id, version_md5 FROM build_resource_config_version_inputs i WHERE i.build_id = $1
+							SELECT build_id, resource_id, version_md5
+							FROM build_resource_config_version_inputs i
+							WHERE i.build_id = $1
 						)
 				) AS agg GROUP BY build_id, resource_id) sp ON sp.build_id = b.id
 				WHERE b.id = $1
 				GROUP BY b.id, b.job_id, b.rerun_of
-			) ON CONFLICT (build_id) DO UPDATE SET outputs = EXCLUDED.outputs RETURNING outputs`, buildID).
+			)
+			ON CONFLICT (build_id) DO UPDATE SET outputs = EXCLUDED.outputs
+			RETURNING outputs
+		`, buildID).
 		Scan(&outputs)
 	if err != nil {
+		tracing.End(span, err)
 		return "", err
 	}
 
+	span.AddEvent(ctx, "build migrated")
+
 	return outputs, nil
 }
 
+type BuildCursor struct {
+	ID      int
+	RerunOf sql.NullInt64
+}
+
+func (cursor BuildCursor) OlderBuilds(idCol string) sq.Sqlizer {
+	if cursor.RerunOf.Valid {
+		return sq.Or{
+			sq.Expr("COALESCE(rerun_of, "+idCol+") < ?", cursor.RerunOf.Int64),
+			sq.And{
+				sq.Eq{"rerun_of": cursor.RerunOf.Int64},
+				sq.Lt{idCol: cursor.ID},
+			},
+		}
+	} else {
+		return sq.Expr("COALESCE(rerun_of, "+idCol+") < ?", cursor.ID)
+	}
+}
+
+func (cursor BuildCursor) NewerBuilds(idCol string) sq.Sqlizer {
+	if cursor.RerunOf.Valid {
+		return sq.Or{
+			sq.Expr("COALESCE(rerun_of, "+idCol+") > ?", cursor.RerunOf.Int64),
+			sq.And{
+				sq.Eq{"rerun_of": cursor.RerunOf.Int64},
+				sq.Gt{idCol: cursor.ID},
+			},
+		}
+	} else {
+		return sq.Expr("COALESCE(rerun_of, "+idCol+") > ?", cursor.ID)
```

Comment:
I think this won't catch the rerun builds if the cursor is on the original build. For example, if we have builds 5.2, 5.1, 5 and the cursor id is 5, any reruns of that build will have a rerun_of of 5 (5.1 and 5.2 will have rerun_of = 5) so they won't be fetched from this query because we only grab anything larger than the cursor ID.

Reply:
> I think this won't catch the rerun builds if the cursor is on the original build. For example, if we have builds 5.2, 5.1, 5 and the cursor id is 5, any reruns of that build will have a rerun_of of 5 (5.1 and 5.2 will have rerun_of = 5) so they won't be fetched from this query because we only grab anything larger than the cursor ID.

:+1: will add a test